### DISCUSSION TITLE
How to generate a conformer from SMILES given coordinates of substructures

### DISCUSSION THREAD

**Discussion by yanze039** [rdkit/rdkit#6590] [Q&A] [answered]
Hi,

I have a molecule with 3D coordinates in `.sdf` file. And I have a chemically modified version of  that molecule in SMILES, which share a common substructure with A. Now I want to generate 3D coordinates of the modified molecule, so that the coordinates of MCS of the two are exactly the same, and the other groups can be arranged properly.

Could someone please give me some hints about how to achieve that?

Thanks!

**Comment by ptosco** [comment]
You may use `GenerateDepictionMatching2DStructure()`.
You'd better normalize the scale of the starting coordinates such that it is in line with what RDKit expects.
In my example I use RDKit for the constrained coordinate generation, but often CoordGen does a better job, even though it may slightly alter your scaffold coordinates to ensure a better final result.
```
from rdkit import Chem
from rdkit.Chem import rdDepictor

rdDepictor.SetPreferCoordGen(False)

tropane = Chem.MolFromMolBlock("""
  Marvin  05050609032D          

  9 10  0  0  0  0            999 V2000
    1.2072    0.6970    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.2072   -0.6970    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    1.6096    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.4024   -0.8785    0.0000 C   0  0  2  0  0  0  0  0  0  0  0  0
   -0.3750   -0.4887    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -0.3750    0.4887    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
    0.4024    0.8785    0.0000 C   0  0  1  0  0  0  0  0  0  0  0  0
    0.0000    0.0000    0.0000 N   0  0  0  0  0  0  0  0  0  0  0  0
    0.8250    0.0000    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
  3  1  1  0  0  0  0
  3  2  1  0  0  0  0
  1  7  1  0  0  0  0
  7  8  1  1  0  0  0
  4  8  1  1  0  0  0
  4  2  1  0  0  0  0
  4  5  1  0  0  0  0
  7  6  1  0  0  0  0
  6  5  1  0  0  0  0
  8  9  1  0  0  0  0
M  END
""")

rdDepictor.NormalizeDepiction(tropane, canonicalize=0)
1.875

tropane
```
![image](https://github.com/rdkit/rdkit/assets/5244385/6a7304dd-67dc-4aad-96ca-945a1935f021)
```
atropine = Chem.MolFromSmiles("CN3[C@H]1CC[C@@H]3C[C@@H](C1)OC(=O)C(CO)c2ccccc2")
rdDepictor.GenerateDepictionMatching2DStructure(atropine, tropane, acceptFailure=False)
((0, 6), (1, 8), (2, 7), (3, 2), (4, 3), (5, 4), (6, 5), (7, 1), (8, 0))

atropine
```
![image](https://github.com/rdkit/rdkit/assets/5244385/eca7f8ea-b6d6-4b46-b557-a797b8b54b05)

**Reply by yanze039** [reply]
Hi Paolo,
Thanks for your detailed explanation. But my case is for 3D coords alignment.
Nevertheless, your code is also good for me. I appreciate it 👍

**Reply by ptosco** [reply]
Ah, sorry, I missed "3D" in your original question.

**Comment by DavidACosgrove** [comment, marked as answer]
For 3D coordinates, you can use ConstrainedEmbed.  See http://rdkit.org/docs/source/rdkit.Chem.AllChem.html for information.  The syntax is similar to Paolo's example, and there's an example in the web page, though the crucial line isn't obvious due to what appears to be a typo in the HTML source.

**Reply by yanze039** [reply]
Hi David,

Thanks for your reply, it helps a lot.

But when I do Embed with coordMap, I got an error like:

```
[01:32:26] Could not triangle bounds smooth molecule.
```

I searched on many websites but I haven't got any intuitive explaination on that. Do you know what does it mean or what can it indicate to my case?

Thanks!

**Reply by DavidACosgrove** [reply]
Hi @Dead-fisher,
Triangle bounds smoothing is a part of the distance geometry algorithm used for creating the conformation.  The effect is that a 3D conformation can't be generated.  I would imagine in this case that it is because the 3D coordinates you have provided for the template can't be married successfully with the rest of the molecule.  There's a little bit of relevant information in this article: https://chempedia.info/info/bounds_smoothing/.  
Dave
